shell/server/proxy.js

```diff
@@ -1118,9 +1118,57 @@ tryProxyRequest = (hostId, req, res) => {
 
     return Promise.resolve(true);
   } else {
-    const isAlreadyOpened = req.headers.cookie && req.headers.cookie.indexOf("sandstorm-sid=") !== -1;
+    const isAlreadyOpened = req.headers.cookie &&
+        req.headers.cookie.indexOf("sandstorm-sid=") !== -1;
+
     return getProxyForHostId(hostId, isAlreadyOpened).then((proxy) => {
       if (proxy) {
+        // Let's do some CSRF defense.
+        const expectedOrigin = PROTOCOL + "//" + req.headers.host;
+        const mainUrl = process.env.ROOT_URL;
+        const origin = req.headers.origin;
+        const referer = req.headers.referer;
+        if (origin) {
+          // If an origin header was provided, then it must be accurate. Note that Chrome and
+          // Safari always send an Origin header on non-GET requests (even same-origin), and
+          // therefore on those browsers this rule will block all CSRF attacks. Unfortunately,
+          // a patch to add this behavior to Firefox has stalled:
+          //     https://bugzilla.mozilla.org/show_bug.cgi?id=446344
+          // I haven't found any information on IE/Edge's behavior.
+          if (origin !== expectedOrigin) {
+            throw new Meteor.Error(403, "Blocked illegal cross-origin request from: " + origin);
+          }
+        } else if (referer) {
+          // Mark that we've seed a Referer header on this host, which indicates that the user's
+          // browser is not suppressing them.
+          proxy.seenReferer = true;
+
+          // Deny requests coming from an external referer, since there is no legitimate use case
+          // for these. Note of course that an attacker can trivially suppress the referer header
+          // in practice, so this check alone does not give us any security benefit if we accept
+          // requests that lack a referrer.
```

Comment:
I recommend the "referer" spelling here, to be consistent with your other comments and the variable names.